Migrate @datx/jsonapi-angular tooling to Angular CLI

fvoska opened the pull request:
There were some issues regarding missing metadata when building DatxModule with rollup so I had to change the tooling to Angular CLI in order to attach the Angular Compiler's medatada to the output bundle.

Next I will work on updating the DatX Angular example repo to use the new setup + update the docs

* [ ] This PR contains a new feature
* [ ] This PR updates an existing feature
* [ ] This PR contains bugfixes
* [ ] This PR contains all the relevant tests
* [ ] This PR creates a breaking change

Please describe the differences between the current and new behavior

DarkoKukovec commented on packages/datx-jsonapi-angular/package.json:
```diff
@@ -1,101 +1,37 @@
-    ]
+    "ng": "ng",
+    "start": "ng serve",
+    "build": "ng build datx-jsonapi-angular && npm run copy-license",
```
license?

fvoska replied:
I wanted the LICENSE file to be part of the NPM package, so I added a script that copies the license file over to dist folder so that it is packed into the tarball when releasing to npm.

DarkoKukovec commented on packages/datx-jsonapi-angular/projects/datx-jsonapi-angular/README.md:
```diff
@@ -0,0 +1,24 @@
+# DatxJsonapiAngular
```
This should be more in line with other readmes (this is used on npm).

DarkoKukovec commented on packages/datx-jsonapi-angular/projects/datx-jsonapi-angular/src/lib/decorateView.ts:
```diff
@@ -21,10 +21,13 @@ export function decorateView<U>(
+        this.__collection
+          .getOne(this.modelType, id, options)
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          .pipe(map((this as any)['__addFromResponse'].bind(this)))
```
what is happening here? why the need for casting?

DarkoKukovec commented on packages/datx-jsonapi-angular/projects/datx-jsonapi-angular/src/lib/services/custom-fetch/custom-fetch.service.ts:
```diff
@@ -19,7 +19,7 @@ export class CustomFetchService {
 
     const requestHeaders = {
-      ...config.defaultFetchOptions.headers,
+      ...config.defaultFetchOptions['headers'],
```
what's the reasoning for this change?

fvoska replied:
When I migrated to angular-cli and TS 4.3, I started getting some more TS errors and this is one of them. I had to go to many places and make this kind of a change for property access. More precisely, this is because `noPropertyAccessFromIndexSignature` is set to `true` in the tsconfig that angular-cli generates. Do you want me to disable that flag and revert the changes that needed to be done to satisfy the rule?

This is the error if I don't make the change:
```
Property 'headers' comes from an index signature, so it must be accessed with ['headers']. ts(4111)
```

DarkoKukovec replied:
I don't know, I don't really like this syntax, and it doesn't make sense to migrate the whole project if it can be disabled.

DarkoKukovec commented on packages/datx-jsonapi-angular/projects/datx-jsonapi-angular/test/general.test.ts:
```diff
@@ -629,16 +629,19 @@ describe('General', () => {
     if (user) {
-      expect(user['selected']).toBeUndefined();
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      expect((user as any)['selected']).toBeUndefined();
```
this doesn't seem ok... looks like all model types are incorrect?

fvoska replied:
Similar as the previous comment, this is due to stricter typechecking after migrating to angular-cli and it's out-of-the-box tsconfig. I will disable some of those checks so that it's more in-line with what we had before

DarkoKukovec replied:
is this promise/observable related? If not, I might want to look into this and try to fix types.

fvoska replied:
I don't think so, I think it was only because of some different tsconfig values. I have reverted them to be more lenient like they were before and removed casting to any (the comment is now outdated)

DarkoKukovec commented on packages/datx-jsonapi-angular/package.json:
```diff
@@ -1,101 +1,37 @@
 {
   "name": "@datx/jsonapi-angular",
   "version": "2.3.1",
-  "description": "DatX mixin for Angular JSON API support",
```
We're missing this whole part in package.json, especially the main and module parts.

fvoska replied:
I've missed re-adding these back in the package.json that angular-cli created from scratch. I will re-add those extra properties that were in the old package.json.

fvoska commented:
@DarkoKukovec I reverted some of the TS-related changes and added a quick readme with basic getting started instructions. I have yet to update the docusaurus docs with same info, perhaps a bit expanded.

kristijanbambir commented on packages/datx-jsonapi-angular/projects/datx-jsonapi-angular/package.json:
```diff
@@ -0,0 +1,14 @@
+  "name": "@datx/jsonapi-angular",
+  "version": "2.3.1",
+  "peerDependencies": {
+    "@angular/common": "^12.2.0",
```
Does this mean that if a project uses Angular 13, npm will complain? If so, then I would set the peer dependencies to >=12.

kristijanbambir commented on packages/datx-jsonapi-angular/package.json:
```diff
@@ -22,80 +22,45 @@
+    "rxjs": "6 || 7"
+  },
+  "devDependencies": {
+    "@angular-devkit/build-angular": "~12.2.13",
```
Perhaps it should be updated to Angular 13.

fvoska replied:
I propose I do that in a separate PR

kristijanbambir commented on packages/datx-jsonapi-angular/projects/datx-jsonapi-angular/package.json:
```diff
@@ -0,0 +1,14 @@
+  "peerDependencies": {
+    "@angular/common": ">=12.2.0",
+    "@angular/core": ">=12.2.0",
+    "@datx/core": "^2.3.1",
```
I missed this before. The peer dependency for all datx packages should be 2.x.

fvoska replied:
I am not 100% sure it will work with e.g. 2.0.0, will check

kristijanbambir replied:
2.x is the same as >=2, so it covers 2.0.0 and upwards.

kristijanbambir replied:
You could put >=2.3.0 then.

fvoska replied:
That's basically the same as `^2.3.0`, should I put that?

kristijanbambir replied:
Ah, you are right 🤦‍♂️  Let's leave it like this then.
